### PR TITLE
Utilities and Arithmetic Rewrite

### PR DESCRIPTION
**Pull request by AaronKutch**
My second big pull request is finally complete. In the end, I realized that this commit has essentially rewritten arithmetic.rs and many utilities, and "multiple commits" in this case would entail each commit being the breaking rewrite of one file. We might as well keep this as one big commit. Trust me when I say that it is a better idea than what my commit history would have looked like.

This includes the renaming of `hi_lo` functions and the zip functions, implementing division and remainder (NOTE: the functions pass both of my special purpose fuzz testers, but there is still a lot of work for me to do in my division function. The only difference now is that we can much more easily work in parallel), fixing up certian docs, deleting `ll.rs` (I might make a new one later), rewriting the testing functions, and a few other things. 

When you compile this you will get two macro use warnings but those can be fixed after your refactor to Rust 2018, which has slightly different macro imports and stuff.

### CODE REVIEW

**Comment by AaronKutch** [comment]
BTW, Git is one of the most nonsensical pieces of software I have ever seen
 - inconsistencies such as -a or -A and -m or --message working or not working on some commands
 - combining multiple commits together seems to work but then trying anything after that corrupts the git history with wierd stuff like staged changes identical to commits that already happened and two commits that have different names but the same changes
 - the way you are supposed to split a commit into multiple commits is very, very confusing and when I tried it it resulted in staged changes trying to add back files that had already been deleted
 - What is even the point of GitHub Desktop? When I first heard of it I thought it was the thing I had been missing, maybe it would allow for visually manipulating a branch graph and quickly selecting parts of a file to be staged. Instead, all it does to improve is to show you the history and diffs better than the command prompt, and doesn't have any commands more advanced than commiting and pushing.
 - It is both vastly simpler, straightforward, and faster to Ctrl-c Ctrl-v files around two local clones with Windows File Explorer, and using my 10 git command cheetsheet to get exactly what I want.

**Comment by AaronKutch** [comment]
Also, when you `cargo doc` this you can also add `document-private-items` now

File: src/apint/arithmetic.rs [hunk 1]
```diff
@@ -407,212 +460,1060 @@ impl ApInt {
...
-//  implementation since a favor in unsigned or signed cannot be decided.
-// ============================================================================
-
-// ============================================================================
-//  Unary arithmetic negation: `std::ops::Add` and `std::ops::AddAssign`
-// ============================================================================
+/// ============================================================================
+///  Standard `ops` trait implementations.
+/// ----------------------------------------------------------------------------
+///  **Note:** These ops will panic if their corresponding functions return an
```
Comment (Robbepop) [inline src/apint/arithmetic.rs]: Good note here! :)

File: src/apint/arithmetic.rs [hunk 2]
```diff
@@ -787,75 +1676,58 @@ mod tests {
...
-                            result.wrapping_lshr_assign(8).unwrap();
-                            if result.clone().resize_to_u8() != u8::MAX {
-                                panic!(
-                                    "\nlhs_nine:{}\nrhs_nine:{}\nresult:{:?}\nzeros_after_one:{}\nnines_before_eight:{}\nnines_after_eight:{}\n",
-                                    lhs_nine, rhs_nine, result,zeros_after_one,nines_before_eight,nines_after_eight
-                                );
+                                continue 'outer
                             }
+                            result.wrapping_lshr_assign(8usize).unwrap();
+                            assert_eq!(result.clone().resize_to_u8(),u8::MAX);
```
Comment (Robbepop) [inline src/apint/arithmetic.rs]: `result.clone()` seems to be a very inefficient operation. So maybe we should use `debug_assert_eq!` here instead of a normal `assert_eq!` to avoid inefficiencies in release code.

Comment (Robbepop) [reply src/apint/arithmetic.rs]: And I think the same could and should be applied to all other applicable cases in this code of course.

File: src/apint/arithmetic.rs [hunk 3]
```diff
@@ -919,68 +1791,374 @@ mod tests {
...
 
+        //TODO: add division by zero testing after error refactoring is finished
+        //use errors::ErrorKind;
         #[test]
         fn simple() {
-            let lhs = ApInt::from(72_i32);
-            let rhs = ApInt::from(12_i32);
-            let result = lhs.into_wrapping_sdiv(&rhs).unwrap();
-            assert_eq!(result, ApInt::from(6_u32));
+            macro_rules! s {
```
Comment (Robbepop) [inline src/apint/arithmetic.rs]: Please add some documentation to this macro (even if its test code) and maybe restructure the parameters so they are more readable. Like columnwise.

Comment (Robbepop) [reply src/apint/arithmetic.rs]: Also I am not sure if you really need a macro here at all. A simple (generic) function should suffice.

File: src/apint/arithmetic.rs [hunk 4]
```diff
@@ -919,68 +1791,374 @@ mod tests {
...
 
+        //NOTE: this test only works if multiplication and a few other functions work
         #[test]
-        fn with_neg() {
-            let lhs = ApInt::from(72_i32);
-            let rhs = ApInt::from(-12_i32);
-            let result = lhs.into_wrapping_sdiv(&rhs).unwrap();
-            assert_eq!(result, ApInt::from(-6_i32));
+        fn complex() {
+            //there are many special case and size optimization paths,
```
Comment (Robbepop) [inline src/apint/arithmetic.rs]: I am always unsure as to how to structure tests for special cases. I like to put every special case in its own test so I have a better overview about what actually is broken whenever I run the tests. However, I do not want to say that this is the best procedure about test structuring and also results in many many test cases.

File: src/apint/constructors.rs [hunk 5]
```diff
@@ -189,6 +190,18 @@ impl ApInt {
         }
     }
 
+    pub fn from_vec_u64(val: Vec<u64>) -> Option<ApInt> {
```
Comment (Robbepop) [inline src/apint/constructors.rs]: I really do not want to have this as public function to be honest. At least we should be very cafeful about this public API addition. Maybe we could stabilize it later for public access. It is too special, reflects too much of the inner working to the outside and the `Option<ApInt>` return type should also be more in par with the rest of the library and return a `Result` instead.

Comment (Robbepop) [reply src/apint/constructors.rs]: Also needs proper documentation.

File: src/apint/utils.rs [hunk 6]
```diff
@@ -62,16 +62,22 @@ pub(crate) enum ZipDataAccess<'a, 'b> {
...
+    Inl(&'a mut Digit, &'b mut Digit),
+    Ext(&'a mut [Digit], &'b mut [Digit])
+}
+
 // ============================================================================
 
 impl Width for ApInt {
     /// Returns the `BitWidth` of this `ApInt`.
-    #[inline]
+    #[inline(always)]
```
Comment (Robbepop) [inline src/apint/utils.rs]: Have you tested for the `#[inline(always)]` attribute? I have already seen reports not to use it too frequently and instead rely on the compiler for the simpler `#[inline]` in the general case.

File: src/apint/utils.rs [hunk 7]
```diff
@@ -209,6 +245,7 @@ impl ApInt {
     }
 
     /// Returns a slice over the `Digit`s of this `ApInt` in little-endian order.
+    #[inline(always)]
```
Comment (Robbepop) [inline src/apint/utils.rs]: Please reconsider if using `#[inline(always)]` in so many cases is really the right option. I can think of cases where the inlining is poisonous to caller code. So I think in the general case it is better to let the compiler decide upon every use case.

File: src/apint/utils.rs [hunk 8]
```diff
@@ -321,33 +366,48 @@ impl ApInt {
...
+    #[inline]    
     pub fn is_one(&self) -> bool {
         match self.access_data() {
             DataAccess::Inl(digit) => digit == Digit::one(),
             DataAccess::Ext(digits) => {
-                let (last, rest) = digits.split_last()
-                    .expect("An `ApInt` always has at least one digit so calling \
-                             `split_last` on a slice of its digits will never \
-                             return `None`.");
+                let (last, rest) = digits.split_last().unwrap_or_else(|| unreachable!());
```
Comment (Robbepop) [inline src/apint/utils.rs]: Maybe we should add a safe API for `split_last` since it cannot fail for all digits as one of their invariants is to never be empty.

File: src/digit.rs [hunk 9]
```diff
@@ -518,25 +612,49 @@ impl BitXor for Digit {
...
+    fn shl(self, rhs: usize) -> Self::Output {
+        Digit(self.repr().wrapping_shl(rhs as u32))
+    }
+}
+
+impl Shr<usize> for Digit {
+    type Output = Digit;
+
+    fn shr(self, rhs: usize) -> Self::Output {
+        Digit(self.repr().wrapping_shr(rhs as u32))
```
Comment (Robbepop) [inline src/digit.rs]: Maybe we should protect against this `as u32` cast via `assert!`.
(in general)

File: src/int.rs [hunk 10]
```diff
@@ -43,7 +43,9 @@ use std::ops::{
 /// 
 /// This very cheaply transformes to and from `ApInt` and `UInt` instances and together with
 /// `UInt` offers a more elegant and higher-level abstraction interface to the lower-level `ApInt`.
-#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(serde_derive, Serialize)]
```
Comment (Robbepop) [inline src/int.rs]: Nice! :)

**Comment by Robbepop** [comment]
Again thank you so much for all the excellent work you have put into this library! :)
I really want to merge this. See my PR review above.

File: src/apint/arithmetic.rs [hunk 11]
```diff
@@ -53,154 +77,184 @@ impl ApInt {
...
+                        }
+                        //no excess bits to clear
+                        return Ok(carry != Digit::zero())
+                    }
+                }
+            }
+        }
     }
 
+    //pub(crate) fn overflowing_sadd_assign()
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: There is a difference between signed and unsigned overflow however. e.x. if we have two four bit `ApInt`s, one of which is 1110 (-2 in the signed interpretation) and another 1110, and we add them, it will make 1100 (-4). There are carries off the end but not numerical overflow

File: Cargo.toml [hunk 12]
```diff
@@ -13,20 +13,21 @@ description = """Arbitrary precision integers library."""
 categories = ["data-structures"]
 
 [dependencies]
-smallvec = "0.6"
-rand = { version = "0.5", optional = true }
-serde = { version = "1.0", optional = true }
-serde_derive = { version = "1.0", optional = true }
+smallvec = { version = "0.6.5", features = ["union"] }
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply Cargo.toml]: Oops I forgot to remove the `union` feature. I did that during benchmarking and I can remove it. Do you also mean to ask why I chose those version numbers? I just chose the latest version I could find. IIRC, smallvec has had a lot of perf improvements in the past months.

File: src/apint/arithmetic.rs [hunk 13]
```diff
@@ -53,154 +77,184 @@ impl ApInt {
...
+
+    /// Add-assigns `rhs` to `self` inplace, and returns a boolean indicating if overflow occured,
+    /// according to the **unsigned** interpretation of overflow.
     /// 
     /// # Errors
     /// 
     /// - If `self` and `rhs` have unmatching bit widths.
-    pub fn into_wrapping_sub(self, rhs: &ApInt) -> Result<ApInt> {
-        try_forward_bin_mut_impl(self, rhs, ApInt::wrapping_sub_assign)
+    pub fn overflowing_uadd_assign(&mut self, rhs: &ApInt) -> Result<bool> {
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: Most of the code there won't be useful anywhere else except if I implement `overflowing_uadd_assign`. I might however do something about the loops in this file. You see, I have barely used any iterators in my code because of the massive amount of special cases, preconditions, postconditions, and other things. There is a very common thing I use that involves

```
                        //pre loop carry setups with some function
                        let (temp, mut carry) = lhs[0].carrying_add(rhs[0]);
                        lhs[0] = temp;
                        for i in 1..(lhs.len() - 1) {
                            //main for body using the carries and the same function
                            let temp = lhs[i].dd().wrapping_add(rhs[i].dd()).wrapping_add(carry.dd());
                            lhs[i] = temp.lo();
                            carry = temp.hi();
                        }
                        //post loop no carry set part
                        let temp = lhs[lhs.len() - 1]
                            .wrapping_add(rhs[lhs.len() - 1])
                            .wrapping_add(carry);
```
And I will probably make some kind of macro or method for `Digit`s for that. Another reason for not using iterators is because I don't entirely trust that the compiler can optimize away certain things, and it is harder to reason about iterators when there is a lot of lazy evaluation, closures, borrowing, and dereferencing flying around.

File: src/apint/arithmetic.rs [hunk 14]
```diff
@@ -407,212 +460,1060 @@ impl ApInt {
...
+/// - any division that will lead to the quotient being zero or one (no allocation)
+/// - division of any integer by small (1 `Digit`) very small (0.5 `Digit`) integers (no allocation)
+/// - division where the number of leading zeros of both arguments are within one `Digit` (less allocation than what long division normally requires)
+/// - during long division, the algorithm may encounter a case from above and will use that instead
+/// - division of medium size (<= 512 bits) integers
+/// 
+/// Currently, algorithms faster than 𝒪(n^2) are not implemented, so large integer division may be
+/// very slow compared to other algorithms.
+impl ApInt {
+    //Note: the invariant of `ApInt`s where unused bits beyond the bit width must be all zero is
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: hmm I thought that the white space after `///` was just there because it makes an even 4 characters. Should I add a space after every `//`?

**Comment by AaronKutch** [comment]
Do I just click `resolve conversation` if I did a quick fix for it? Also, I forgot about how to change the commits in the pull request without closing it.

File: src/apint/utils.rs (hunk 8 above)
Comment (AaronKutch) [reply src/apint/utils.rs]: Should I make an issue for that? There are a lot of other `.expect`s around the library that might be eliminated this way.

File: src/apint/arithmetic.rs [hunk 15]
```diff
@@ -407,212 +460,1060 @@ impl ApInt {
...
+                        }
+                    }
+                    b1
+                }) {$le_branch} else {$gt_branch}
+            };
+        }
+
+        //assigns `$sum + $sub` to `$target`,
+        //and zeros out `$sum` except for it sets `$sum[0]` to `$val`
+        macro_rules! special0 {
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: I don't know of a better way to do it. I think these macros (except maybe the `uge` macro) are only going to be used in this one function, and not using the macro would entail error prone copy paste anyway.

File: src/apint/arithmetic.rs [hunk 16]
```diff
@@ -407,212 +460,1060 @@ impl ApInt {
...
+                        let temp = duo[0].wrapping_divrem(div[0]);
+                        duo[0] = temp.0;
+                        div[0] = temp.1;
+                        return true
+                    }
+                }
+            },
+            None => return false,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: Yep that's what I think you could do next. I will just deal with what is inside the functions, and you can deal with moving them relative to each other in the module system.

File: src/apint/arithmetic.rs (hunk 1 above)
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: When I compile the with `cargo doc`, I don't see the large blocks not attached to `impl`s or functions or am I blind? Should we document this somewhere else because the trait impls are way down there in the docs?

File: src/apint/arithmetic.rs (hunk 2 above)
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: Do you realize that is testing code? It has to be `assert_eq` because I compile with `cargo test --release`. Also, I should note that running all the tests requires `--release` now because of the fuzz testers, or else it will take a long time. At the very bottom of `megafuzz` there are commented out lines that take something like an hour to run so I only run them whenever I am going to issue a pull request that modifies arithmetic functions. The `.clone` I am using everywhere in the tests are needed because I need to test one thing against many things.

File: src/apint/arithmetic.rs (hunk 3 above)
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: When I add the error checking and potentially other things, this macro will really save on lines and effort. I think it is already worth it and in fact it helped me find a bug.

File: src/apint/arithmetic.rs (hunk 4 above)
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: There are thousands of special cases I am testing there. Some of this is actually brute force fuzzing that I designed specifically for `ApInt`s and the functions on them. I didn't encounter any really crazy edge or corner cases, but these tests are what give me a lot of confidence in my functions

File: src/apint/constructors.rs (hunk 5 above)
Comment (AaronKutch) [reply src/apint/constructors.rs]: You're right. I will make it private for now

File: src/apint/utils.rs (hunk 6 above)
Comment (AaronKutch) [reply src/apint/utils.rs]: I overreached again now I realize. I searched it and found some interesting things I didn't know before. I am using it on internal functions that are used only a few times or on one-liners that probably only need `inline`. I think I will look at the perf a little more and remove some instances.

File: src/apint/utils.rs (hunk 8 above)
Comment (Robbepop) [reply src/apint/utils.rs]: I think it doesn't matter if it also is part of this giantic PR. So just add it if you want but please with proper doc etc. ;)

File: src/apint/arithmetic.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Robbepop) [reply src/apint/arithmetic.rs]: I have made many performance tests about iterators because I also didn't believe that the Rust compiler is able to optimize them as well as other constructs but in the end iterator and especially high-level programming in general was nearly always the way that the compiler optimized best in release mode. Of course in debug mode it oftentimes was the slowest but we do not mind for debug mode and care only for release mode when looking at performance.

[the remark this answers is not in the document]
Comment (Robbepop) [reply src/apint/arithmetic.rs]: The logic behind this is: The compiler has an easier life determining the logic behind high-level code and a much harder time doing this for low-level code constructs.

File: src/apint/utils.rs (hunk 6 above)
Comment (Robbepop) [reply src/apint/utils.rs]: As far as I have tested one should really only ever use `#[inline(always)]` for private one-liners that have been tested extensively about performance aspects for the certain use-case since `#[inline(always)]` is more or less pretty restrictive in its behaviour.

File: src/apint/arithmetic.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Robbepop) [reply src/apint/arithmetic.rs]: `ApInt` as I have designed it is unrelated to signedness. So only operations defined on it are either signed or unsigned operations. So there is basically no difference between `uadd` and `sadd` in the general case and if we have the overflowing version that tells us if there is an overflow bit we do not need one without that in my opinion. I hope I understood this correctly.

[the remark this answers is not in the document]
Comment (Robbepop) [reply src/apint/arithmetic.rs]: Note that `Int` and `Uint` are existent to provide a signed and unsigned version and API for the underlying `ApInt` exactly for the reason stated above.

[the remark this answers is not in the document]
Comment (AaronKutch) [reply src/apint/arithmetic.rs]: I still think there should be two functions, because testing for signed overflow would involve
```
let overflow = lhs.overflowing_add_assign(rhs).unwrap();
signed_overflow = (lhs.is_negative() == rhs.is_negative()) && overflow;
```
which I would put in a overflowing_sadd_assign

**Comment by AaronKutch** [comment]
It appears that the pull request automatically updates when I force the commit

**Comment by AaronKutch** [comment]
There were no perf regressions from removing the `inline(always)`. The two overflowing functions and the `from_vec_u64` I will fix or rewrite in the future, I have made them `pub(crate)` for now.

**Comment by Robbepop** [comment]
> It appears that the pull request automatically updates when I force the commit

That is at least how I also always update my PR commit history.

**Comment by AaronKutch** [comment]
Are there any other problems to fix before accepting the PR?

**Comment by AaronKutch** [comment]
I also haven't seen you comment in the refactoring issue. Have you seen it or did the Github notification go in the email spam folder like it does to me sometimes?

**Comment by Robbepop** [comment]
I am so sorry. I will answer you tomorrow when I find time for it. Hope that is okay. I think at least to me it looks like there are no major problems that should prevent this from being merged. ;)

**Comment by Robbepop** [comment]
I think this is ready for merging.
We can continue working out the details in other PRs I think.
